Refactor tags

dometto opened the pull request:
This refactor shaves about 5 seconds of load time for a page with a ridiculous amount (1000) of tags (~25 seconds vs. ~30), and (hopefully) improves readability and maintainability. To further improve performance I would like to eventually collapse the search for files and pages so we don't traverse the repository twice, and of course implement caching of the repo tree map. But that requires refactoring the `Page` and `File` classes, and implementing the idea for a `Gollum::IO` class...

Some breaking changes:
  * syntax for link attributes changed from `image.jpg|frame|alt=Bla` to `image.jpg|frame, alt=Bla`
  * image links and file links and are now also reversed if `@markup.reverse_links?` (previously only for page links and external links). This is because [reversal is now handled at the start of the link processing rather than in each method separately](https://github.com/gollum/gollum-lib/pull/269/files#diff-9c66db1000504e8fd86544f23961803eR100).
  * a page link to `waa.md` would *first* be interpreted as a relative link, and if that didn't find a page, it would be interpreted as an absolute link (to `/waa.md`). So in cases of broken links the repository was traversed twice. I've disabled this, so an absolute link now requires an explicit `/`. However, this change doesn't break any tests.

dometto commented:
Todo: update inline docs for tag filter.

bartkamphorst commented on lib/gollum-lib/filter/tags.rb:
```diff
@@ -4,7 +4,8 @@
   # Extract all tags into the tagmap and replace with placeholders.
   def extract(data)
-    return data if @markup.format == :txt || @markup.format == :asciidoc
+    return data if @markup.skip_tags?
```
Expresses the intent much clearer, nice. 👍

bartkamphorst commented on lib/gollum-lib/filter/tags.rb:
```diff
@@ -73,21 +74,31 @@ def is_preformatted?(node)
+  end
+
+  def process_link_tag(link_part, extra)
+    process_file_link_tag(link_part, extra) || process_page_link_tag(link_part, extra)
```
This means that all tags are processed as file links initially. Is there any way to avoid this?

dometto replied:
Yes, that's what bothered me originally. I would like to just walk through the repo searching for the requested paths, returning `Page` objects if the found object has a valid page extensions and `File` objects otherwise. Then `process_link_tag` should choose what to with each link based on what kind of object it gets back (if any).

dometto replied:
But I think that may require a new approach to `Page` and `File` objects altogether. Perhaps have `BlobEntry` superclass of which `Page` and `File` are both subclasses, with a common find-and-construct-from-path method. Or just have the `BlobEntry` class with a `#file?` and `#page?` method?

bartkamphorst commented on lib/gollum-lib/filter/tags.rb:
```diff
@@ -98,10 +109,9 @@ def process_tag(tag)
   #
   def process_include_tag(tag)
-    return unless /^include:/.match(tag)
+    return html_error('Cannot process include directive: no page name given') if tag.length <= 8
```
What is this magic number 8?

dometto replied:
:) Point take 👍

bartkamphorst commented on lib/gollum-lib/filter/tags.rb:
```diff
@@ -98,10 +109,9 @@ def process_tag(tag)
     page_name          = tag[8..-1]
-    resolved_page_name = ::File.expand_path(page_name, "/"+@markup.dir)
-
+    resolved_page_name = ::File.expand_path(page_name, "#{::File::SEPARATOR}#{@markup.dir}")
```
Nice (helpful going forward on Windows). I guess it doesn't make sense to work with Pathname objects here, right?

dometto replied:
Not sure. I tried putting a `::File.join` here but that broke some tests. The point here is to prevent people from including `/etc/password`. `File.expand_path` here never returns a path lower in the directory tree than `@markup.dir`. Does `Pathname` have a similar method?

bartkamphorst replied:
I'm not sure, but we could just build a Pathname object from the string that `expand_path` comes up with. The bigger question is whether we need something more than a string here. I guess time will tell. ;)

dometto replied:
Right. If we want to rebuild the API for searching/retrieving objects from the repo to use `Pathname`, we should also do it here.

bartkamphorst commented on lib/gollum-lib/filter/tags.rb:
```diff
@@ -120,23 +130,19 @@ def process_include_tag(tag)
       path = name
-    elsif name =~ /.+(jpg|png|gif|svg|bmp)$/i
+    elsif (file = @markup.find_file(name))
+      path = ::File.join @markup.wiki.base_path, file.path
```
Stylistically I'd like to see the join call have parentheses.

bartkamphorst commented on lib/gollum-lib/filter/tags.rb:
```diff
@@ -246,12 +246,7 @@ def process_external_link_tag(tag)
-    path = parts[1] && parts[1].strip
+  def process_file_link_tag(name, path)
     if path && file = @markup.find_file(path)
       path = ::File.join @markup.wiki.base_path, file.path
```
Parentheses.

bartkamphorst replied:
This methods returns HTML. Is there any way to refactor that so that gollum (and not gollum-lib) wraps `path` and `name` in the appropriate html tags?

dometto commented:
@bartkamphorst: incorporated your stylistic suggestions, @bartkamphorst, and refactored further.

I don't see how we can easily get rid of returning HTML inside the tag filter, since the purpose of that filter (and other filters) *is* basically to generate HTML on top of what the markup parser generates. However, I've tried to at least cleanly separate the tag-parsing logic from the html-generation logic. Hardcoded HTML is now only present in `#generate_link` and `#generate_image` and one or two helper methods they call.

One way to get the hardcoded HTML out of `gollum-lib` would perhaps be to move `#generate_link` and `#generate_image` to `gollum`, and somehow pass them on to the tag filter as a `Proc` from there on? But that would mean that the filter by itself would not be fully functional 'out of the box'.

dometto commented:
And as you suggested, `process_file_link_tag` and `process_page_link_tag` should be collapsed into a single method. But I would propose to merge this now, and to take that further step once we have refactored the `Page` and `File` classes.

dometto commented:
Another change in behavior affected by this PR is that `File` links without a description now work, whereas in `master` they seem to result in a broken `img` for some reason. The latter clearly seems like a bug, so I'll add a regression test.

dometto commented:
Re-implemented the tag skipping logic in the `Markup` class as you suggested, reverted the changes to the `Filter` class. I think this is nice and clean now, thanks for the review!

bartkamphorst commented on lib/gollum-lib/markup.rb:
```diff
@@ -126,7 +126,7 @@ def render_default(data, format=:markdown, name='render_default.md')
       @name   = name
 
-      chain = [:YAML, :PlainText, :Emoji, :TOC, :RemoteCode, :Code, :Sanitize, :PlantUML, :Tags, :Render]
+      chain = [:YAML, :PlainText, :Emoji, :TOC, :RemoteCode, :Code, :Sanitize, :PlantUML, :Tags, :Render].reject {|filter| skip_filter?(filter)}
```
Why here?

dometto replied:
You mean why in the `render_default` method? It's used in the tests at least, so filter skipping should work there too, I thought.

bartkamphorst replied:
Because you also call `.reject` on line 180.

dometto replied:
Right, but the `render` and `render_default` methods are enitrely separate from each other, so the rejecting needs to happen twice. But I now notice that `render_default` is called just once in `gollum-lib` and `gollum` taken together, namely in a test. If I modify that test we can just remove this method alltogeter. :+1:

bartkamphorst replied:
👍

dometto replied:
`render_default` was being used to allow use of the `Markup` class without a wiki. I can see some usecases for that, but I don't know we should consider them within the legitimate scope of the gollum project. Perhaps in the interest of general code sanity, we can remove it?

bartkamphorst replied:
Go for it. 😄 People can always come knocking if they miss this functionality in 5.x.

bartkamphorst replied:
Might be good to add to the 5.x changelog though!
